drivers/SmartThings/zigbee-window-treatment/src/hanssem/init.lua

```diff
@@ -0,0 +1,259 @@
+	local PayloadBody = generic_body.GenericBody(strSeqNum .. DpId .. Type .. LenOfValue .. Value)
+	local MsgBody = ZigbeeZcl.ZclMessageBody({zcl_header = zclh, zcl_body = PayloadBody})
+	local TxMsg = Messages.ZigbeeMessageTx({address_header = addrh, body = MsgBody})
+	Log.debug(string.format(" ###########>>>>>>>>>>>> Send DpId : %d   Value : %d", string.byte(DpId,1), string.unpack(">I"..string.len(Value),Value)))
+	device:send(TxMsg)
+end
+
+----------------- Functions to emit a capability event -------------------
+
+local function getLatestLevel(device)
+	local ret = device:get_latest_state("main", capabilities.windowShadeLevel.ID, capabilities.windowShadeLevel.shadeLevel.NAME)
+	if ret == nil then ret = 0 end
+	Log.debug("Level State : " .. ret)
+	return ret
+end
+
+local function emit_event_movement_status(device, target_level)
+	local ret = false
+	local current_level = getLatestLevel(device)
+	if current_level ~= nil and current_level ~= target_level then
+		if current_level > target_level then
+			device:emit_event(capabilities.windowShade.windowShade.closing())			
+		elseif (current_level < target_level) then
+			device:emit_event(capabilities.windowShade.windowShade.opening())			
+		end
+		ret =  true
+	end
+	return ret
+end
+
+local function emit_event_final_position(device, feedback_level)
+	local window_shade_val
+	if type(feedback_level) ~= "number" then
+		window_shade_val = "unknown"
+		feedback_level = 50
+	elseif feedback_level == 0 then
+		window_shade_val = "closed"
+	elseif feedback_level == 100 then 
+		window_shade_val = "open"
+	elseif feedback_level > 0 and feedback_level < 100 then
+		window_shade_val = "partially open"
+	else
+		window_shade_val = "unknown"
+		feedback_level = 50
+	end
+	device:emit_event(capabilities.windowShadeLevel.shadeLevel(feedback_level))	
+	device:emit_event(capabilities.windowShade.windowShade(window_shade_val))
+end
+
+-------------------- Capability Handlers -----------------------
+
+local DP_ID_CONTROL = "\x01"
+local DP_ID_SET_POSITION= "\x02"
+local DP_ID_RESET_DIRECTION = "\x05"
+local DP_VAL_OPEN = "\x00"
+local DP_VAL_PAUSE = "\x01"
+local DP_VAL_CLOSE = "\x02"
+local DP_VAL_DIRECT = "\x00"
+local DP_VAL_REVERSE = "\x01"
+
+local function OpenHandler(driver, device, capability_command)
+	local level = getLatestLevel(device)	
+	if level == 100 then
+		device:emit_event(capabilities.windowShade.windowShade.open())
+	else
+		SendCommand(device, DP_ID_CONTROL, DP_TYPE_ENUM, DP_VAL_OPEN)
+	end
+end
+
+local function CloseHandler(driver, device, capability_command)
+	local level = getLatestLevel(device)	
+	if level == 0 then
+		device:emit_event(capabilities.windowShade.windowShade.closed())
+	else
+		SendCommand(device, DP_ID_CONTROL, DP_TYPE_ENUM, DP_VAL_CLOSE)
+	end
+end
+
+local function PauseHandler(driver, device, capability_command)
+	local ShadeState = device:get_latest_state("main", capabilities.windowShade.ID, capabilities.windowShade.windowShade.NAME)
```

Comment:
why not use your `getLatestLevel` function?

Reply:
The getLatestLevel function returns the windowShadeLevel, but the ShadeState requires the windowShade capability.

Reply:
sorry, misread